microraiden/microraiden/test/test_http_client.py

```diff
@@ -1,47 +1,212 @@
 from requests.exceptions import SSLError
 
+from microraiden import HTTPHeaders
 from microraiden import DefaultHTTPClient
-from microraiden.test.utils.client import close_channel_cooperatively, patch_on_http_response
+from microraiden.test.utils.client import patch_on_http_response
 from microraiden.test.utils.disable_ssl_check import disable_ssl_check
 
 
 def check_response(response: bytes):
     assert response and response.decode().strip() == '"HI I AM A DOGGO"'
 
 
-def test_cheating_client(
-        doggo_proxy,
-        default_http_client: DefaultHTTPClient
+def test_full_cycle_success(
+        default_http_client: DefaultHTTPClient,
+        api_endpoint_address: str,
+        token_contract_address,
+        channel_manager_contract_address,
+        receiver_address
+):
+    default_http_client.initial_deposit = lambda x: x
+
+    with requests_mock.mock() as server_mock:
+        headers1 = Munch()
+        headers1.token_address = token_contract_address
+        headers1.contract_address = channel_manager_contract_address
+        headers1.receiver_address = receiver_address
+        headers1.price = '7'
+
+        headers2 = Munch()
+        headers2.cost = '7'
+
+        headers1 = HTTPHeaders.serialize(headers1)
+        headers2 = HTTPHeaders.serialize(headers2)
+
+        url = 'http://{}/something'.format(api_endpoint_address)
+        server_mock.get(url, [
+            {'status_code': 402, 'headers': headers1},
+            {'status_code': 200, 'headers': headers2, 'text': 'success'}
+        ])
+        resource = default_http_client.run('something')
+
+    # First cycle, request price.
+    request = server_mock.request_history[0]
+    assert request.path == '/something'
+    assert request.method == 'GET'
+    assert request.headers['RDN-Contract-Address'] == channel_manager_contract_address
+
+    # Second cycle, pay price.
+    request = server_mock.request_history[1]
+    assert request.path == '/something'
+    assert request.method == 'GET'
+    assert request.headers['RDN-Contract-Address'] == channel_manager_contract_address
+    assert request.headers['RDN-Balance'] == '7'
+    assert default_http_client.channel.balance == 7
+    balance_sig_hex = encode_hex(default_http_client.channel.balance_sig)
+    assert request.headers['RDN-Balance-Signature'] == balance_sig_hex
+    assert default_http_client.channel.balance_sig
+    assert resource == b'success'
+
+
+def test_full_cycle_adapt_balance(
+        default_http_client: DefaultHTTPClient,
+        api_endpoint_address: str,
+        token_contract_address,
+        channel_manager_contract_address,
+        receiver_address
 ):
+    # Simulate a lost balance signature.
+    client = default_http_client.client
+    channel = client.get_suitable_channel(receiver_address, 10, initial_deposit=lambda x: 2 * x)
+    channel.create_transfer(3)
+    lost_balance_sig = channel.balance_sig
+    channel.balance = 0
+
+    with requests_mock.mock() as server_mock:
```

Comment:
Good idea but we still need to have client-proxy integration tests :)

Reply:
Sooon. Although many are already in there.